src/olmo_core/nn/lm_head.py

```diff
@@ -350,7 +374,13 @@ def apply_tp(
                 parallelize_plan=SequenceParallel(),
             )
 
-        if self.loss_implementation != LMLossImplementation.fused_linear:
+        if self.loss_implementation == LMLossImplementation.fused_linear:
+            parallelize_module(
+                module=self.w_out,
+                device_mesh=tp_mesh,
+                parallelize_plan=SequenceParallel(),
+            )
+        else:
```

Comment:
This is the one change needed to fix issue (2), everything else was the fix issue (1).